service/accessor.go

```diff
@@ -96,3 +100,237 @@ func (ua *UpdatableAccessor) Update(a Accessor, err error) {
 	ua.current = a
 	ua.lock.Unlock()
 }
+
+type RouteTraffic interface {
+	Route(instance string) error
+}
+
+type emptyRouter struct {
+}
+
+func (r emptyRouter) Route(instance string) error {
+	return nil
+}
+
+func DefaultTrafficRouter() RouteTraffic {
+	return emptyRouter{}
+}
+
+type AccessorQueue interface {
+	Order([]string) []string
+}
+
+type emptyQueue struct {
+	r *rand.Rand
+}
+
+func DefaultOrder() AccessorQueue {
+	return emptyQueue{}
+}
+
+func (r emptyQueue) Order(keys []string) []string {
+	return keys
+}
+
+type AccessorValue struct {
+	Accessor Accessor
+	Err      error
+}
+
+type LayeredAccessor struct {
+	lock sync.RWMutex
+
+	router        RouteTraffic
+	accessorQueue AccessorQueue
+
+	err     error
+	primary Accessor
+
+	failover map[string]AccessorValue
+}
+
+// SetRouter will update teh router, which will determine if the accessor should return the instance or fail
+func (la *LayeredAccessor) SetRouter(router RouteTraffic) {
```

Comment:
As an example of mutability:  This method mutates the `LayeredAccessor` type, forcing you to use a mutex.

If a `NewLayeredAccessor` constructor was provided, the router can be supplied up front and thus be immutable for the life of the `LayeredAccessor` object.